### ISSUE TITLE
Request new configuration remotely

### ISSUE THREAD

**Issue by laurijamsa** [ruuvi/ruuvi.gateway_esp.c#455]
During the configuration wizard, one should be able to define a remote URL from where the Gateway would poll new configuration every now and then.

This is related to the issue https://github.com/ruuvi/ruuvi.gateway_esp.c/issues/454

The new config could be either a bin file or a json. If we'll have an easy to use config creation web page that outputs .bin as suggested in the issue https://github.com/ruuvi/ruuvi.gateway_esp.c/issues/454, maybe that would be the way to go.

**Comment by laurijamsa** [comment]
> It would perhaps also be interesting to be able to supply the remote URL via a DHCP-option?

It is actually already possible to configure remotely your Gateways one by one like this: https://github.com/ruuvi/ruuvi.gateway_esp.c/issues/435

**Comment by olemd** [comment]
I don't see that, it allows enabling web-access via ethernet, but not auto-config?

If I was unclear, it would be nice with a DHCP-option. Something like "option-ruuvi https://server/config.json" that the gateway could pick up via DHCP and automatically use config.json - completely hands-off configuration.

**Comment by laurijamsa** [comment]
> I don't see that, it allows enabling web-access via ethernet, but not auto-config?

Yes, correct. No auto configuration feature exist yet. Thanks for the clarification.

**Comment by Scrin** [comment]
The [dev documentation](https://docs.ruuvi.com/) is a bit lacking regarding all this, but either case it would be nice if the configuration could be updated either in a "pull" (ie. polling) or "push" (which is technically possible now, but undocumented), ideally in the same format for consistency.

Additionally, especially for "pull setups", it would be convenient to have a (documented) method of checking the current config or a checksum of it, in order to be able to monitor which gateway(s) are already running an updated config and whether some gateways get a new config unintentionally.

One simple solution to this (which I would prefer) would be a `ruuvigw_config_checksum <crc32 of the config>` metric to the /metrics endpoint: this way secrets are not exposed, but by knowing your config versions you can determine which config is currently present based on the crc. (crc32 because prometheus exposition format values are 64bit floats, and thus crc64 and "proper" hash functions would produce too large numbers)

**Comment by ojousima** [comment]
We're going to have a minor version update for the configuration, let's pack all configuration-related issues to 1.11.0. See also #456

**Comment by TheSomeMan** [comment]
@ojousima I added CRC32 to the metrics endpoint, but CRC32 is calculated for the full configuration, while "/ruuvi.json" endpoint does not contain passwords, so it's not possible for the user to calculate/compare CRC32.

**Comment by ojousima** [comment]
Good point, thanks. Let's look into it next week and decide what's the best way to handle it

**Comment by TheSomeMan** [comment]
> If I was unclear, it would be nice with a DHCP-option. Something like "option-ruuvi https://server/config.json" that the gateway could pick up via DHCP and automatically use config.json - completely hands-off configuration.

@olemd which DHCP-option do you propose to use for this?

**Comment by olemd** [comment]
It would seem like option 43 (Vendor Specific Information) is the only reasonable choice, unless you want to talk to iana.org and get a dedicated ruuvi-option.

https://www.rfc-editor.org/rfc/rfc2132.html#section-8 See section 8.4. This would allow a dhcp server to supply an option with something like "ruuvi-bootstrap:https://server/config.json"

It would probably make sense to use something shorter than "ruuvi-boostrap:" though, since the option payload is limited in size.

(When I first suggested this I somewhat naively thought you could supply named options in dhcp, as in literally "option-ruuvi". Turns out that's not possible.)

**Comment by DG12** [comment]
Of course there are also the officially unused options:
https://datatracker.ietf.org/doc/html/rfc3679 and 
https://www.iana.org/assignments/bootp-dhcp-parameters/bootp-dhcp-parameters.xhtml
84, 96,102-107 ,110, 115, 126,127, 149,160, etc

**Comment by ojousima** [comment]
I opened a new issue for DHCP configuration at #532 . If we go forward with it, we would schedule it to the version v1.14.X where we support more third-party integrations. 

My main concern about it is in security, is there a way someone could inject DHCP option to hijack the Gateways of common users?

**Comment by olemd** [comment]
The officially unused options should probably not be used. That's just waiting for a collision with a new official option.

Obviously just accepting a config.json via a dhcp-option would allow a malicious dhcp-server to hijack a gateway, I'd suggest that the dhcp-option should only be used when booting in reset-mode. So if there is no config (except whatever is default), and the dhcp-server provides a url for config.json then use it. But if the gateway has a non-standard config, then just ignore the dhcp-option.

**Comment by TheSomeMan** [comment]
To test it:

    Need to create an initial version of gw_cfg.json (use example from https://app.gitbook.com/o/-LsPfwGmkUWZFBz3qkE9/s/-LsPjfnhNj3v37AZtF-d/data-formats/gateway-configuration)
    Configure WiFi in "wifi_sta_config"
    Configure "remote_cfg_use", "remote_cfg_url", "remote_cfg_refresh_interval_minutes"
    Run the local HTTP server and make this gw_cfg.json accessible
    Configure gateway - on the Remote Configuration Download enable this feature and press Download button
    Change something in gw_cfg.json (for example, enable MQTT)
    Wait for a specified period ("remote_cfg_refresh_interval_minutes")
    Check that the new configuration was downloaded and applied

**Comment by ojousima** [comment]
I cannot download config over HTTPS. 

This happens both with and without basic authentication. CFG is at https://storage.ruuvi.com/D869/gw_cfg.json, user "ruuvi" password "tester"

Remote server Apache2, using HTTPS over Let's Encrypt cert and automatic redirect from HTTP to HTTPS.

```
I (2148847) http_server: [http_server] cb_on_http_download_json_data: buf_size=333
E (2148855) http_server: [http_server] /Users/otsojousimaa/git/ruuvi.gateway_esp.c/main/http_server_cb.c:268 {cb_on_http_download_json_data}: Got HTTP error 301: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>301 Moved Permanently</title>
</head><body>
<h1>Moved Permanently</h1>
<p>The document has moved <a href="https://storage.ruuvi.com/D869/gw_cfg.json">here</a>.</p>
<hr>
<address>Apache/2.4.25 (Debian) Server at storage.ruuvi.com Port 80</address>
</body></html>

I (2148901) http: [http_server] HTTP_EVENT_DISCONNECTED
I (2148908) http: [http_server] HTTP_EVENT_ON_FINISH
I (2148912) esp-tls: [http_server] hostname 'storage.ruuvi.com' resolved to 167.99.134.205
I (2149382) esp-x509-crt-bundle: Certificate validated
I (2150872) http: [http_server] HTTP_EVENT_HEADER_SENT
I (2151075) http: [http_server] HTTP_EVENT_ON_HEADER, key=Date, value=Wed, 15 Jun 2022 16:45:04 GMT
I (2151075) http: [http_server] HTTP_EVENT_ON_HEADER, key=Server, value=Apache/2.4.25 (Debian)
I (2151083) http: [http_server] HTTP_EVENT_ON_HEADER, key=Last-Modified, value=Wed, 15 Jun 2022 16:20:25 GMT
I (2151094) http: [http_server] HTTP_EVENT_ON_HEADER, key=ETag, value="77f-5e17ee6160d1a"
I (2151103) http: [http_server] HTTP_EVENT_ON_HEADER, key=Accept-Ranges, value=bytes
I (2151111) http: [http_server] HTTP_EVENT_ON_HEADER, key=Content-Length, value=1919
I (2151119) http: [http_server] HTTP_EVENT_ON_HEADER, key=Content-Type, value=application/json
I (2151134) http_server: [http_server] cb_on_http_download_json_data: buf_size=1919
E (2151137) http_server: [http_server] /Users/otsojousimaa/git/ruuvi.gateway_esp.c/main/http_server_cb.c:258 {cb_on_http_download_json_data}: Error occurred while downloading
I (2151153) http: [http_server] HTTP_EVENT_ON_FINISH
I (2151159) http: [http_server] HTTP_EVENT_DISCONNECTED
E (2151168) http_server: [http_server] /Users/otsojousimaa/git/ruuvi.gateway_esp.c/main/http_server_cb.c:375 {http_download_json}: http_download returned NULL buffer
I (2151180) http_server: [http_server] http_download_json: completed within 2750 ticks
E (2151189) http_server: [http_server] /Users/otsojousimaa/git/ruuvi.gateway_esp.c/main/http_server_cb.c:875 {http_server_gw_cfg_download_and_update}: Download gw_cfg: failed, http_resp_code=400
W (2151207) http_server: [http_server] Response: status 400 (Bad Request)
I (2151373) http_server: [http_server] Request from 192.168.168.232 to 192.168.168.51: GET /status.json
I (2151376) http_server: [http_server] Json resp: code=200, content:
{"ssid":"ruuvitag","ip":"192.168.168.51","netmask":"255.255.255.0","gw":"192.168.168.175","dhcp":"192.168.168.175","urc":0,"lan":1}

I (2151392) http_server: [http_server] Response: OK
```

<img width="795" alt="Näyttökuva 2022-6-15 kello 19 43 14" src="https://user-images.githubusercontent.com/2360368/173881682-fb155f2c-23ef-47ed-8573-0fd8bf08cbe4.png">

**Comment by TheSomeMan** [comment]
It seems that redirection is not working for some reason, but if you explicitly specify "https://" protocol in "Base URL", then it should work. Try this: "https://storage.ruuvi.com/D869"

**Comment by ojousima** [comment]
This is a bit different issue, but if there is a problem with downloaded JSON the Gateway UI reports "503"

<img width="741" alt="image" src="https://user-images.githubusercontent.com/2360368/174972785-63a6b9df-d79a-44a4-8642-f3507c53a195.png">

Logs have
```
I (264139) http_server: [http_server] Request from 10.10.0.2 to 10.10.0.1: POST /gw_cfg_download
I (264140) http_server: [http_server] POST /gw_cfg_download
I (264144) http: [http_server] http_download: https://storage.ruuvi.com/D869/gw_cfg.json
I (264155) esp-tls: [http_server] hostname 'storage.ruuvi.com' resolved to 167.99.134.205
I (264371) esp-x509-crt-bundle: Certificate validated
I (265578) http: [http_server] HTTP_EVENT_HEADER_SENT
I (265631) http: [http_server] HTTP_EVENT_ON_HEADER, key=Date, value=Wed, 22 Jun 2022 07:30:41 GMT
I (265631) http: [http_server] HTTP_EVENT_ON_HEADER, key=Server, value=Apache/2.4.25 (Debian)
I (265639) http: [http_server] HTTP_EVENT_ON_HEADER, key=Last-Modified, value=Wed, 15 Jun 2022 16:20:25 GMT
I (265649) http: [http_server] HTTP_EVENT_ON_HEADER, key=ETag, value="77f-5e17ee6160d1a"
I (265658) http: [http_server] HTTP_EVENT_ON_HEADER, key=Accept-Ranges, value=bytes
I (265667) http: [http_server] HTTP_EVENT_ON_HEADER, key=Content-Length, value=1919
I (265675) http: [http_server] HTTP_EVENT_ON_HEADER, key=Content-Type, value=application/json
I (265689) http_server: [http_server] cb_on_http_download_json_data: buf_size=1919
I (265693) http: [http_server] HTTP_EVENT_ON_FINISH
I (265698) http: [http_server] HTTP_EVENT_DISCONNECTED
I (265707) http_server: [http_server] http_download_json: completed within 1563 ticks
I (265712) http_server: [http_server] Download gw_cfg: successfully completed
E (265720) gw_cfg: [http_server] /Users/otsojousimaa/git/ruuvi.gateway_esp.c/main/gw_cfg_json_parse.c:950 {gw_cfg_json_parse}: Failed to parse gw_cfg.json
E (265734) http_server: [http_server] /Users/otsojousimaa/git/ruuvi.gateway_esp.c/main/http_server_cb.c:897 {http_server_gw_cfg_download_and_update}: Failed to parse gw_cfg.json or no memory
```

**Comment by ojousima** [comment]
Another thing that came to my mind while testing https://github.com/ruuvi/ruuvi.gateway_esp.c/issues/454, it would be good if the configuration URL tried unique URL first and common URL then. The Unique URL could be the Gateway MAC address and common URL gw_cfg.json. 

For example in my case I'd drop the folder D869 and use https://storage.ruuvi.com as base URL. 

The Gateway would then try `https://storage.ruuvi.com/F9158F92D869.json` first and `https://storage.ruuvi.com/gw_cfg.json` second. This would help in cases where client has Gateways in different locations and needs different WiFi credentials or other per-installation settings.

**Comment by TheSomeMan** [comment]
> I cannot download config over HTTPS.

Fixed here [link](https://github.com/ruuvi/ruuvi.gateway_esp.c/issues/555)

**Comment by TheSomeMan** [comment]
> However, some settings are not understood from configuration JSON: At least remote configuration refresh interval and WiFi are not updated.

"remote_cfg_refresh_interval_minutes" is the period between readings of the configuration from the server. In gw_cfg.json on the server it should not be zero. I rechecked how it works and found no problems. Here is the log:
```
I (64626) http: [http_server] http_download: http://192.168.1.119:8000/my123/gw_cfg.json
I (64632) TRANS_TCP: [http_server] hostname '192.168.1.119' resolved to 192.168.1.119
I (64696) http: [http_server] HTTP_EVENT_HEADER_SENT
I (64748) http: [http_server] HTTP_EVENT_ON_HEADER, key=Server, value=SimpleHTTP/0.6 Python/3.8.10
I (64748) http: [http_server] HTTP_EVENT_ON_HEADER, key=Date, value=Wed, 29 Jun 2022 16:47:40 GMT
I (64758) http: [http_server] HTTP_EVENT_ON_HEADER, key=Content-type, value=application/json
I (64765) http: [http_server] HTTP_EVENT_ON_HEADER, key=Content-Length, value=1546
I (64773) http: [http_server] HTTP_EVENT_ON_HEADER, key=Last-Modified, value=Wed, 29 Jun 2022 16:47:09 GMT
I (64784) http_server: [http_server] cb_on_http_download_json_data: buf_size=1546
I (64792) http: [http_server] HTTP_EVENT_ON_FINISH
I (64797) http: [http_server] HTTP_EVENT_DISCONNECTED
I (64805) http: [http_server] HTTP_EVENT_DISCONNECTED
I (64809) http_server: [http_server] http_download_json: completed within 183 ticks
I (64817) http_server: [http_server] Download gw_cfg: successfully completed
I (64827) gw_cfg: [http_server] Read Gateway SETTINGS from remote server:
I (64833) gw_cfg: [http_server] config: device_info: WiFi AP SSID / Hostname: 
I (64840) gw_cfg: [http_server] config: device_info: ESP32 fw ver: 
I (64847) gw_cfg: [http_server] config: device_info: ESP32 WiFi MAC ADDR: 
I (64855) gw_cfg: [http_server] config: device_info: ESP32 Eth MAC ADDR: 
I (64862) gw_cfg: [http_server] config: device_info: NRF52 fw ver: 
I (64869) gw_cfg: [http_server] config: device_info: NRF52 MAC ADDR: 
I (64876) gw_cfg: [http_server] config: device_info: NRF52 DEVICE ID: 
W (64887) gw_cfg: [http_server] Can't find key 'ntp_use' in config-json, use default value: 'true'
W (64893) gw_cfg: [http_server] Can't find key 'ntp_use_dhcp' in config-json, use default value: 'false'
W (64903) gw_cfg: [http_server] Can't find key 'ntp_server1' in config-json
W (64911) gw_cfg: [http_server] Can't find key 'ntp_server2' in config-json
W (64919) gw_cfg: [http_server] Can't find key 'ntp_server3' in config-json
W (64926) gw_cfg: [http_server] Can't find key 'ntp_server4' in config-json
I (64935) gw_cfg: [http_server] config: use remote cfg: 1
I (64940) gw_cfg: [http_server] config: remote cfg: URL: http://192.168.1.119:8000/my123/gw_cfg.json
I (64950) gw_cfg: [http_server] config: remote cfg: auth_type: basic
I (64957) gw_cfg: [http_server] config: remote cfg: auth user: user1
I (64964) gw_cfg: [http_server] config: remote cfg: auth pass: ********
I (64972) gw_cfg: [http_server] config: remote cfg: refresh_interval_minutes: 2
I (64979) gw_cfg: [http_server] config: use http: 1
I (64985) gw_cfg: [http_server] config: http url: https://network.ruuvi.com/record
I (64993) gw_cfg: [http_server] config: http user: 
I (64998) gw_cfg: [http_server] config: http pass: ********
I (65005) gw_cfg: [http_server] config: use http_stat: 1
I (65011) gw_cfg: [http_server] config: http_stat url: https://network.ruuvi.com/status
I (65019) gw_cfg: [http_server] config: http_stat user: 
I (65025) gw_cfg: [http_server] config: http_stat pass: ********
I (65032) gw_cfg: [http_server] config: use mqtt: 1
I (65037) gw_cfg: [http_server] config: mqtt transport: TCP
I (65044) gw_cfg: [http_server] config: mqtt server: test.mosquitto.org
I (65051) gw_cfg: [http_server] config: mqtt port: 1883
I (65057) gw_cfg: [http_server] config: mqtt prefix: ruuvi/C8:25:2D:8E:9C:2C/
I (65065) gw_cfg: [http_server] config: mqtt client id: C8:25:2D:8E:9C:2C
I (65072) gw_cfg: [http_server] config: mqtt user: 
I (65078) gw_cfg: [http_server] config: mqtt password: ********
I (65084) gw_cfg: [http_server] config: LAN auth type: lan_auth_allow
I (65091) gw_cfg: [http_server] config: LAN auth user: 
I (65097) gw_cfg: [http_server] config: LAN auth pass: ********
I (65104) gw_cfg: [http_server] config: LAN auth API key: ********
I (65111) gw_cfg: [http_server] config: Auto update cycle: manual
I (65118) gw_cfg: [http_server] config: Auto update weekdays_bitmask: 0x7f
I (65125) gw_cfg: [http_server] config: Auto update interval: 00:00..24:00
I (65133) gw_cfg: [http_server] config: Auto update TZ: UTC+3
I (65139) gw_cfg: [http_server] config: NTP: Use: yes
I (65145) gw_cfg: [http_server] config: NTP: Use DHCP: no
I (65151) gw_cfg: [http_server] config: NTP: Server1: time.google.com
I (65158) gw_cfg: [http_server] config: NTP: Server2: time.cloudflare.com
I (65166) gw_cfg: [http_server] config: NTP: Server3: time.nist.gov
I (65172) gw_cfg: [http_server] config: NTP: Server4: pool.ntp.org
I (65179) gw_cfg: [http_server] config: use company id filter: 1
I (65186) gw_cfg: [http_server] config: company id: 0x0499
I (65192) gw_cfg: [http_server] config: use scan coded phy: 0
I (65199) gw_cfg: [http_server] config: use scan 1mbit/phy: 1
I (65205) gw_cfg: [http_server] config: use scan extended payload: 1
I (65212) gw_cfg: [http_server] config: use scan channel 37: 1
I (65219) gw_cfg: [http_server] config: use scan channel 38: 1
I (65225) gw_cfg: [http_server] config: use scan channel 39: 1
I (65232) gw_cfg: [http_server] config: coordinates: 
I (65237) gw_cfg: [http_server] config: Use eth: no
I (65243) gw_cfg: [http_server] config: eth: use DHCP: yes
I (65249) gw_cfg: [http_server] config: eth: static IP: 
I (65255) gw_cfg: [http_server] config: eth: netmask: 
I (65261) gw_cfg: [http_server] config: eth: GW: 
I (65266) gw_cfg: [http_server] config: eth: DNS1: 
I (65272) gw_cfg: [http_server] config: eth: DNS2: 
I (65278) gw_cfg: [http_server] config: wifi_ap_config: SSID: RuuviGateway9C2C
I (65285) gw_cfg: [http_server] config: wifi_ap_config: password: 
I (65292) gw_cfg: [http_server] config: wifi_ap_config: ssid_len: 0
I (65299) gw_cfg: [http_server] config: wifi_ap_config: channel: 1
I (65306) gw_cfg: [http_server] config: wifi_ap_config: auth_mode: OPEN
I (65313) gw_cfg: [http_server] config: wifi_ap_config: ssid_hidden: 0
I (65320) gw_cfg: [http_server] config: wifi_ap_config: max_connections: 4
I (65328) gw_cfg: [http_server] config: wifi_ap_config: beacon_interval: 100
I (65336) gw_cfg: [http_server] config: wifi_ap_settings: bandwidth: 20MHz
I (65343) gw_cfg: [http_server] config: wifi_ap_settings: IP: 10.10.0.1
I (65351) gw_cfg: [http_server] config: wifi_ap_settings: GW: 10.10.0.1
I (65358) gw_cfg: [http_server] config: wifi_ap_settings: Netmask: 255.255.255.0
I (65366) gw_cfg: [http_server] config: wifi_sta_config: SSID:'WIFI768_2.4G', password:'********'
I (65375) gw_cfg: [http_server] config: wifi_sta_config: scan_method: Fast
I (65383) gw_cfg: [http_server] config: wifi_sta_config: Use BSSID: false
I (65390) gw_cfg: [http_server] config: wifi_sta_config: Channel: 0
I (65397) gw_cfg: [http_server] config: wifi_sta_config: Listen interval: 0
I (65405) gw_cfg: [http_server] config: wifi_sta_config: Sort method: by RSSI
I (65413) gw_cfg: [http_server] config: wifi_sta_config: Fast scan: min RSSI: 0
I (65421) gw_cfg: [http_server] config: wifi_sta_config: Fast scan: weakest auth mode: 0
I (65430) gw_cfg: [http_server] config: wifi_sta_config: Protected Management Frame: Capable: false
I (65439) gw_cfg: [http_server] config: wifi_sta_config: Protected Management Frame: Required: false
I (65449) gw_cfg: [http_server] config: wifi_sta_settings: Power save: NONE
I (65457) gw_cfg: [http_server] config: wifi_sta_settings: Use Static IP: no
I (65465) ruuvi_gateway: [http_server] ruuvi_cb_on_change_cfg: settings_save_to_flash
I (65508) ADV_POST_TASK: [adv_post_task] Got ADV_POST_SIG_GW_CFG_CHANGED_RUUVI
I (65514) settings: [http_server] Save config to NVS: successfully updated
I (65519) ruuvi_gateway: [http_server] ruuvi_cb_on_change_cfg: http_server_set_auth: lan_auth_allow
I (65529) http_server: [http_server] Ruuvi configuration in gw_cfg.json differs from the current settings, need to restart services
I (65541) ruuvi_gateway: [http_server] Reading of the configuration from the remote server is active, period: 2 minutes
```

P.S. Current file https://storage.ruuvi.com/D869/gw_cfg.json is not valid, it contains unnecessary framing square brackets.

**Comment by TheSomeMan** [comment]
> Another thing that came to my mind while testing #454, it would be good if the configuration URL tried unique URL first and common URL then. The Unique URL could be the Gateway MAC address and common URL gw_cfg.json.

Implemented here: https://github.com/ruuvi/ruuvi.gateway_esp.c/pull/557
